// @adminjs/cli/lib/commands/create/types.js
export var AdminJSAdapter;
(function (AdminJSAdapter) {
    AdminJSAdapter["SQL"] = "sql";
    AdminJSAdapter["TypeORM"] = "typeorm";
    AdminJSAdapter["Sequelize"] = "sequelize";
    AdminJSAdapter["MikroORM"] = "mikroorm";
    AdminJSAdapter["Prisma"] = "prisma";
    AdminJSAdapter["Objection"] = "objection";
    AdminJSAdapter["Mongoose"] = "mongoose";
})(AdminJSAdapter || (AdminJSAdapter = {}));
export var AdminJSPlugin;
(function (AdminJSPlugin) {
    AdminJSPlugin["Express"] = "express";
    AdminJSPlugin["Hapi"] = "hapi";
    AdminJSPlugin["Koa"] = "koa";
    AdminJSPlugin["NestJS"] = "nestjs";
    AdminJSPlugin["Fastify"] = "fastify";
})(AdminJSPlugin || (AdminJSPlugin = {}));
export var PackageManager;
(function (PackageManager) {
    PackageManager["NPM"] = "npm";
    PackageManager["Yarn"] = "yarn";
})(PackageManager || (PackageManager = {}));
export var DatabaseDriver;
(function (DatabaseDriver) {
    DatabaseDriver["MariaDB"] = "mariadb";
    DatabaseDriver["MSSQL"] = "mssql";
    DatabaseDriver["MySQL"] = "mysql";
    DatabaseDriver["PostgreSQL"] = "postgres";
    DatabaseDriver["Oracle"] = "oracle";
    DatabaseDriver["SQLite"] = "sqlite";
    DatabaseDriver["SQLJS"] = "sqljs";
    DatabaseDriver["MongoDB"] = "mongodb";
})(DatabaseDriver || (DatabaseDriver = {}));
